### ISSUE TITLE
libOpenGL.so.0 not deployed when running with -bundle-non-qt-libs

### ISSUE THREAD

**Issue by alemuntoni** [probonopd/linuxdeployqt#486]
Hi!

When switching from ubuntu 16.04 to 18.04 (with the last linuxdeployqt build), I noticed that libOpenGL.so.0 is not deployed and needs to be installed manually when running a deployed package. This was not necessary when deploying from 16.04. Is this intended?

This is an example workflow that builds and deploys the software, and then runs some tests in a clean machine: https://github.com/alemuntoni/PyMeshLab/runs/2490522243?check_suite_focus=true#step:4:66

As you can see, the deploy works as expected, but tests fail with the error:
```
ImportError: libOpenGL.so.0: cannot open shared object file: No such file or directory
``` 

Installing the package `libopengl0` before running the tests solves the issue ([commit](https://github.com/alemuntoni/PyMeshLab/commit/35b0cd9c008b378f2f44386b0ebf4cf1c4e46522) and [workflow](https://github.com/alemuntoni/PyMeshLab/actions/runs/806429165)), but this was not necessary when deploying from 16.04.

[This](https://github.com/alemuntoni/PyMeshLab/blob/master/scripts/Linux/2_deploy.sh) is the deploy script that runs `linuxdeployqt` with the `-bundle-non-qt-libs` parameter.

**Comment by hcorion** [comment]
See https://github.com/AppImage/pkg2appimage/issues/477
If your building using CMake do the following `set(OpenGL_GL_PREFERENCE LEGACY)`
@TheAssassin this issue is going to keep popping up as more people move their build systems to 18.04

**Comment by probonopd** [comment]
> Installing the package libopengl0 before running the tests solves the issue

There is nothing that linuxdeployqt can do if a library needed by the application has not been installed on the system before linuxdeployqt is run. Hence closing here.

**Comment by alemuntoni** [comment]
@probonopd 

> > Installing the package libopengl0 before running the tests solves the issue
> 
> There is nothing that `linuxdeployqt` can do if a library needed by the application has not been installed on the system before linuxdeployqt is run. Hence closing here.

@probonopd "Installing the package libopengl0 before running the tests solves the issue" was referred on my (already deployed) application tests, not on `linuxdeployqt`.

**Comment by probonopd** [comment]
What in the AppDir is needing `libOpenGL.so.0`?

**Comment by alemuntoni** [comment]
In our apps (meshlab and pymeshlab) almost all targets (binaries, libraries and plugins) need opengl.
In our cmake configuration we use
```
find_package(OpenGL REQUIRED)
```
and the output is 
```
Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so
```

I tried the solution suggested by @hcorion, but it seems not to work. 
Before building the project, I am installing the following packages (I don't know if I'm missing something):
```
libglut3-dev freeglut3-dev mesa-common-dev libglu1-mesa-dev
```

But as far as I understand, `libOpenGL.so` has been deliberately excluded on #152, right?

**Comment by hcorion** [comment]
Hmmm, setting that worked for my project, I'd be curious to see the output of lddtree, I wonder if Qt opengl is trying to include it as well

**Comment by alemuntoni** [comment]
Small update: after making sure that the cmake config of meshlab had the [CMP0072](https://cmake.org/cmake/help/git-stage/policy/CMP0072.html) set to old everywhere, in order to use legacy opengl, there was still something that was linking `libOpenGL.so`. I am not sure if it is related to the fact that we need the target `OpenGL::GLU`...
Anyway, I was not able to get rid of `libOpenGL.so` in favour of `libGL.so`.

In the meantime, I am manually copying `libOpenGL.so` in the AppDir, as it is done in [this commit](https://github.com/hcorion/rpcs3/commit/a70284629597715656e0bcc35bc7939096da228d) of RPCS3. This workaround seems to work, at least on Ubuntu.

It would be great to know if `libOpenGL.so` on Ubuntu 18.04 really is a hardware-dependent library, and therefore if it would be safe to get it bundled in an AppDir.

**Comment by hcorion** [comment]
We removed the manual addition in RPCS3 because it was causing issues with distros that already had a libOpenGL.so, it is unfortunately a distro-dependent library and will cause breakages. That's very strange that it just won't link against libGL.

As a better solution, I would recommend doing what the PCSX2 AppImage is doing https://github.com/PCSX2/pcsx2/pull/4103/files#diff-dab8bc68effd20fe4cf10daad595a9f919536bec53e406549cc331aa38f27392R13-R15
It's a bit of an ugly hack, but it'll work better on all distros, instead of just Ubuntu.

**Comment by alemuntoni** [comment]
Oh, ok.. Thanks @hcorion. I'll then revert that commit :D 

Your hack seems fine just for AppImages, though.. In my case, the appdir is also "transformed" in a python library (with python bindings), and I am not sure that it could be done.. Apparently also Qt6 [is having similar issues with libopengl.so](https://bugreports.qt.io/browse/QTBUG-89754). 

I think that I'll end up with a warning to run `sudo apt install libopengl0` before running the library on ubuntu...

**Comment by hcorion** [comment]
That's probably for the best for now until the AppImage folks come up with a better solution 👍

**Comment by arximboldi** [comment]
I am struggling with this very same issue.  I really do not like the solution of asking people to install `libopengl0` since this is the whole point of AppImage :)

In my case, I think the problem is that the Qt I am installing (from prebuilt binaries elsewhere) already links to `libOpenGL`. I am considering just using `patchelf` to change the reference `libOpenGl.so.0` for something else. The question is: what is this something else? Should it be `libGL.so`? Or what else?

**Comment by hcorion** [comment]
Did you try setting `set(OpenGL_GL_PREFERENCE LEGACY)` in CMake? This is what we did to fix the AppImage on RPCS3.

**Comment by arximboldi** [comment]
Yes, sorry for the late reply but actually in the end I realized the issue was not coming from the prebuilt Qt binaries but from a CMake submodule I was linking against OpenGL, and I managed to fix it that way. Thanks a lot for the quick response also @hcorion !
